Sources/Differ/Diff+UIKit.swift

```diff
@@ -253,16 +253,17 @@ public extension UICollectionView {
     ///   - oldData:            Data which reflects the previous state of `UICollectionView`
     ///   - newData:            Data which reflects the current state of `UICollectionView`
     ///   - indexPathTransform: Closure which transforms zero-based `IndexPath` to desired  `IndexPath`
+    ///   - updateData:         Closure to be called immediately before performing updates, giving you a chance to correctly update data source
     ///   - completion:         Closure to be executed when the animation completes
     func animateItemChanges<T: Collection>(
         oldData: T,
         newData: T,
         indexPathTransform: @escaping (IndexPath) -> IndexPath = { $0 },
-        updateData: ((T) -> Void)? = nil,
+        updateData: () -> Void,
```

Comment:
```suggestion
        updateData: () -> Void = {},
```

I think it's worth considering a default parameter value here and in the other methods affected by this PR. There's potential for some users to not need it, and this change forces them to use an empty closure here. Providing a default value allows them to elide the parameter from the call site. What do you think?

Reply:
I'm 100% behind making this closure non-optional, but I'm also not 100% convinced that we _shouldn't_ be passing `newData` through to this closure. It avoids potential issues around unintended captures for this closure.

Reply:
1) As for capturing, we'll end up capturing our `dataSource` anyway in order to be able to do 
`apply(diff, updateData: { self.data = newData })`. Furthermore, `performBatchUpdates` is not `@escaping` meaning there won't be any capturing,  because such closure is guaranteed to return before the calling method returns.

2) As for `updateData` default parameter - it might look convenient for those who don't use it. But, not using it is implicitly creates a possibility to crash. `UICollectionView` API explicitly says that one should update `dataSource` within `performBatchUpdates` call to avoid potential issues (e.g. unexpected crashes). So I believe it would be better to enforce consumer to use that closure for his own safety. Though, I'm not 100% sure that it is responsibility to protect consumer, but I'd personally ensure that framework did everything  to prevent issues. Ultimately that is up to you to decide whether this change looks reasonable for you or not.

Reply:
To argue that this is not merely stylistic changes I'd like to add some context  to it. In the project I'm currently working on, there is a modular architecture, which among other benefits isolates data from UI layer. With that system, I was quite confused as for why `Differ` suddenly put a hard requirement on passing a `data` up to UI layer, that would require certain violation of existing architecture, so before doing so I've investigated why that `data` became required in the first place. That's how I came up with removing it altogether. So now UI will only notify Data layer to update dataSource at appropriate time.

Reply:
> `UICollectionView` API explicitly says that one should update `dataSource` within `performBatchUpdates` call to avoid potential issues (e.g. unexpected crashes). So I believe it would be better to enforce consumer to use that closure for his own safety. 

Perhaps, we can reflect it additionally in the documentation of that closure parameter to clarify why a consumer should update `dataSource` within `updateData`.

Reply:
That's a good idea - would you be willing to draft something to add to the existing documentation in this PR? This all makes sense, as does your rationalisation of the change.

Reply:
Yes :) will do.